### PR TITLE
[#103] Common check: module names in PascalCase

### PR DESCRIPTION
**Pull request by jiegillet**
Solves #103.

First try with the analyzer, so I'm not sure I got everything right. Is there something else I should run ti check the PR besides `mix test`?

I noticed at the end that Credo calls this check `ModuleName` and I called it `ModulePascalCase`. Is that OK?

Sorry I couldn't review the previous PR, family time comes first :)

### CODE REVIEW

File: test/elixir_analyzer/exercise_test/common_checks_test.exs [hunk 1]
```diff
@@ -127,4 +128,46 @@ defmodule ElixirAnalyzer.ExerciseTest.CommonChecksTest do
...
+          defmodule SomeSubModule do
+          end
+
+          defmodule SomeOtherSubModule do
+          end
+        end
+      ]
+    end
+
+    test_exercise_analysis "reports a module attribute that doesn't use snake_case",
```
Comment (angelikatyborska) [inline test/elixir_analyzer/exercise_test/common_checks_test.exs:153]:
```suggestion
    test_exercise_analysis "reports a module attribute that doesn't use PascalCase",
```

**Review by angelikatyborska — APPROVED** [review]
> First try with the analyzer, so I'm not sure I got everything right.

It's perfect, thank you 💙 the only missing part is a corresponding PR with the actual comment, to be added here: https://github.com/exercism/website-copy/tree/main/analyzer-comments/elixir/solution

> Is there something else I should run ti check the PR besides mix test?

Nope, I just need to allow the CI check to run and that's it :)

> I noticed at the end that Credo calls this check ModuleName and I called it ModulePascalCase. Is that OK?

Ok!

> Sorry I couldn't review the previous PR, family time comes first :)

Of course, no worries

**Comment by jiegillet** [comment]
Wonderful. I'll do the PR on the other side.

File: test/elixir_analyzer/exercise_test/common_checks/module_pascal_case_test.exs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+              %Comment{
+                type: :actionable,
+                comment: Constants.solution_module_pascal_case(),
+                params: %{
+                  expected: "FactorialSubmodule",
+                  actual: "factorial_submodule"
+                }
+              }}
+           ]
+  end
```
Comment (angelikatyborska) [inline test/elixir_analyzer/exercise_test/common_checks/module_pascal_case_test.exs:99]: On second thought... let's add some more tests for corner cases (that fail atm):
```suggestion
  end
  
  test "reports a module name with many segments" do
    code =
      quote do
        defmodule MyLibrary.Math_ops.Factorial do
        end
      end

    assert ModulePascalCase.run(code) == [
             {:fail,
               %Comment{
                 type: :actionable,
                 comment: Constants.solution_module_pascal_case(),
                 params: %{
                   expected: "MyLibrary.MathOps.Factorial",
                   actual: "MyLibrary.Math_ops.Factorial"
                 }
               }}
           ]
  end

  test "reports an inner module name with many segments" do
    code =
      quote do
        defmodule MyLibrary do
          defmodule Math_ops.Factorial do
          end
        end
      end

    assert ModulePascalCase.run(code) == [
             {:fail,
               %Comment{
                 type: :actionable,
                 comment: Constants.solution_module_pascal_case(),
                 params: %{
                   expected: "MathOps.Factorial",
                   actual: "Math_ops.Factorial"
                 }
               }}
           ]
  end
```

Comment (jiegillet) [reply test/elixir_analyzer/exercise_test/common_checks/module_pascal_case_test.exs:99]: Thank you, I was careless.

I did not consider segmented module names AND I did not realize that module names cannot start with a lowercase character to begin with, so many of my tests weren't valid. 

It would have been found out if the `imperfect_solution`'s tests had been ran as well. Is there a way to add that test?

**Comment by neenjaw** [comment]
Looks good. just one nitpick.  Since we have several people actively working in this repo, can you make sure to put your name on the issue you are working on?

**Review by neenjaw — APPROVED** [review]

File: lib/elixir_analyzer/constants.ex [hunk 3]
```diff
@@ -19,6 +19,7 @@ defmodule ElixirAnalyzer.Constants do
     solution_use_specification: "elixir.solution.use_specification",
     solution_raise_fn_clause_error: "elixir.solution.raise_fn_clause_error",
     solution_module_attribute_name_snake_case: "elixir.solution.module_attribute_name_snake_case",
+    solution_module_pascal_case: "elixir.solution.module_name_pascal_case",
```
Comment (neenjaw) [inline lib/elixir_analyzer/constants.ex:22]: There is a file name mismatch here.  In website/copy, this is the file name:

```suggestion
    solution_module_pascal_case: "elixir.solution.module_pascal_case",
```

**Review by neenjaw — CHANGES_REQUESTED** [review]
There is a file name mismatch causing a test to fail

**Comment by jiegillet** [comment]
Thank you! 
Yes. I will call the issue in advance from now on.

**Comment by neenjaw** [comment]
I think a test file is now failing because it expects the old comment string

**Comment by jiegillet** [comment]
Yes, I just noticed. I probably shouldn't commit code from GitHub directly :)

**Review by neenjaw — APPROVED** [review]
